Subject: [PATCH] fix: apply https://go.dev/cl/600296
---
Index: src/runtime/cgo/gcc_stack_unix.c
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/runtime/cgo/gcc_stack_unix.c b/src/runtime/cgo/gcc_stack_unix.c
--- a/src/runtime/cgo/gcc_stack_unix.c	(revision 5f0e1c070c4854f2e2137b3812f03db5376baa0c)
+++ b/src/runtime/cgo/gcc_stack_unix.c	(revision b510c5d15f43712df1df7cce09abbc2b8ce147b5)
@@ -18,6 +18,9 @@
 	void *addr;
 	size_t size;
 
+    // Needed before pthread_getattr_np, too, since before glibc 2.32
+    // it did not call pthread_attr_init in all cases (see #65625).
+    pthread_attr_init(&attr);
 #if defined(__GLIBC__) || defined(__BIONIC__) || (defined(__sun) && !defined(__illumos__))
 	// pthread_getattr_np is a GNU extension supported in glibc.
 	// Solaris is not glibc but does support pthread_getattr_np
@@ -25,7 +28,6 @@
 	pthread_getattr_np(pthread_self(), &attr);  // GNU extension
 	pthread_attr_getstack(&attr, &addr, &size); // low address
 #elif defined(__illumos__)
-	pthread_attr_init(&attr);
 	pthread_attr_get_np(pthread_self(), &attr);
 	pthread_attr_getstack(&attr, &addr, &size); // low address
 #else
