Subject: [PATCH] [release-branch.go1.22] runtime: reserve 4kB for system stack on windows-386
---
Index: src/runtime/stack.go
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/runtime/stack.go b/src/runtime/stack.go
--- a/src/runtime/stack.go	(revision bf937668e64105fe4f7652f3d6c4b39fb2e45400)
+++ b/src/runtime/stack.go	(revision ec0d75a8abb3c9a34b6dea44aedcb4168f097c33)
@@ -69,7 +69,7 @@
 	// to each stack below the usual guard area for OS-specific
 	// purposes like signal handling. Used on Windows, Plan 9,
 	// and iOS because they do not use a separate stack.
-	stackSystem = goos.IsWindows*512*goarch.PtrSize + goos.IsPlan9*512 + goos.IsIos*goarch.IsArm64*1024
+	stackSystem = goos.IsWindows*4096 + goos.IsPlan9*512 + goos.IsIos*goarch.IsArm64*1024
 
 	// The minimum size of stack used by Go code
 	stackMin = 2048
@@ -1301,7 +1301,7 @@
 }
 
 // startingStackSize is the amount of stack that new goroutines start with.
-// It is a power of 2, and between _FixedStack and maxstacksize, inclusive.
+// It is a power of 2, and between fixedStack and maxstacksize, inclusive.
 // startingStackSize is updated every GC by tracking the average size of
 // stacks scanned during the GC.
 var startingStackSize uint32 = fixedStack
