Subject: [PATCH] [release-branch.go1.23] cmd/go/internal/modfetch: do not trust server to send all tags in shallow fetch
---
Index: src/cmd/go/internal/modfetch/codehost/git.go
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/cmd/go/internal/modfetch/codehost/git.go b/src/cmd/go/internal/modfetch/codehost/git.go
--- a/src/cmd/go/internal/modfetch/codehost/git.go	(revision fa5190f00b6621649cb5eee39c02d2d0ac5cf26a)
+++ b/src/cmd/go/internal/modfetch/codehost/git.go	(revision adbf4faba61791285e6b37e52382d2946c184b98)
@@ -662,7 +662,21 @@
 			}
 		}
 	}
-	sort.Strings(info.Tags)
+
+	// Git 2.47.1 does not send the tags during shallow clone anymore
+	// (perhaps the exact version that changed behavior is an earlier one),
+	// so we have to also add tags from the refs list we fetched with ls-remote.
+	if refs, err := r.loadRefs(ctx); err == nil {
+		for ref, h := range refs {
+			if h == hash {
+				if tag, found := strings.CutPrefix(ref, "refs/tags/"); found {
+					info.Tags = append(info.Tags, tag)
+				}
+			}
+		}
+	}
+	slices.Sort(info.Tags)
+	info.Tags = slices.Compact(info.Tags)
 
 	// Used hash as info.Version above.
 	// Use caller's suggested version if it appears in the tag list
